[api] Add QgsProject::crs3D

nyalldawson opened the pull request:
Returns the CRS to use for the project when transforming 3D data, or when z/elevation value handling is important.

The returned CRS will take into account verticalCrs(), e.g. by returning a compound CRS consisting of crs() + verticalCrs().

This method may still return a 2D CRS, e.g in the case that crs() is a 2D CRS and no verticalCrs() has been set for the project. I.E. It is NOT guaranteed that the returned CRS will actually be a 3D CRS, but rather it is guaranteed that the returned CRS is **ALWAYS** the most appropriate CRS to use when handling 3D data.

Refs discussion in https://github.com/qgis/QGIS/pull/56791, https://github.com/qgis/QGIS-Enhancement-Proposals/issues/267

nyalldawson committed:
[api] Add QgsProject::crs3D

Returns the CRS to use for the project when transforming 3D data,
or when z/elevation value handling is important.

The returned CRS will take into account verticalCrs(), e.g. by
returning a compound CRS consisting of crs() + verticalCrs().

This method may still return a 2D CRS, e.g in the case that crs()
is a 2D CRS and no verticalCrs() has been set for the project.
I.E. It is NOT guaranteed that the returned CRS will actually
be a 3D CRS, but rather it is guaranteed that the returned CRS
is **ALWAYS** the most appropriate CRS to use when handling 3D data.

nyalldawson commented:
@rouault @wonder-sk as previously discussed

uclaros commented on src/core/project/qgsproject.h:
```diff
@@ -370,13 +370,36 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
+     * Returns the CRS to use for the project when transforming 3D data, or when z/elevation
+     * value handling is important.
+     *
+     * The returned CRS will take into account verticalCrs(), e.g. by returning a compound
+     * CRS consisting of crs() + verticalCrs(). This method may still return a 2D CRS, e.g in the
+     * case that crs() is a 2D CRS and no verticalCrs() has been set for the project.
```
Maybe it is worth noting how one can verify if the returned crs is actually 3d or 2d?

nyalldawson replied:
Ok, done

wonder-sk commented:
@nyalldawson is this going to work if 3D Geographic (e.g. EPSG:4979) or Geocentric (e.g. EPSG:4978) CRS is set as the project CRS? Or is support for that out of scope for now?

nyalldawson committed:
Expand documentation

nyalldawson committed:
Report proj error when creating compound crs fails

nyalldawson committed:
Improve error reporting in QgsProject.setVerticalCrs

Report proj error when compound crs cannot be created

nyalldawson committed:
Don't permit user set vertical crs for project when crs is geographic 3d
or geocentric

Following https://docs.ogc.org/as/18-005r5/18-005r5.html#34 it is not
a valid combination to create a compound crs from these types

nyalldawson commented:
@wonder-sk I've applied some more related refinements here

nyalldawson committed:
Fix warning

nyalldawson committed:
Make test more stable

rouault commented on src/core/project/qgsproject.cpp:
```diff
@@ -1563,6 +1597,53 @@ void QgsProject::releaseHandlesToProjectArchive()
+      case Qgis::CrsType::Unknown:
+      case Qgis::CrsType::Geodetic:
+      case Qgis::CrsType::Geographic2d:
+      case Qgis::CrsType::Projected:
```
some Projected CRS might already be 3D. You should probably check the axis count

nyalldawson replied:
@rouault 
Thanks!

Does [458b1fd](https://github.com/qgis/QGIS/pull/57107/commits/458b1fde397cabd1f268c58ad248836c0b602e5c) and https://github.com/qgis/QGIS/pull/57107/commits/6773e679c1d195cc0aeb9a4ada75fc53449af853 address this correctly?

rouault replied:
yes, looks good

nyalldawson replied:
Thanks @rouault !

nyalldawson committed:
Add QgsProjUtils and QgsCoordinateReferenceSystem ::hasVerticalAxis

Can be used to test if a crs contains a vertical axis

nyalldawson committed:
Handle projected 3d crs in project vertical/3d crs logic
